### PR TITLE
Real Initialization

### PR DESCRIPTION
**Pull request by haroldcarr**
- Impl/Handle
  - delete fake initialization
  - write real initialization
  - connect real initialization to the system model via `mkSysInitAndHandlers`
  - prove equivalence of `Either` and `EitherD` initialization

- Impl/Handle/InitProperties
  - initialization contract

- Impl/Handle/Properties
  - update `step-init` proofs to use contract defined in `InitProperties`

- Impl/Consensus/ConsensusProvider
  - receive ValidatorSigner as an arg (rather than choosing it inside)
    - changed Haskell in the same way

- Impl/IO/OBM/Start
  - receive ValidatorSigner as an arg (rather than choosing it inside)
    - changed Haskell in the same way

- Impl/OBM/Init
  - deleted redundant file

- Impl/Properties/Common
  - renamed some variables for consistency
  - used holes for unused stuff

- Impl/Properties/VotesOnce
  - renamed some variables for consistency (e.g., `step-init _ uni` (not `ini`)
  - used new required parameter `ini`, when calling `invariantsCorrect`

- ImplShared/Consensus/Types
  - updated BootstrapInfo record
    - to contain the fields really needed for bootstrap
    - to remove `bootstrapQC` (use `bsiLIWS` instead)

- ImplShared/Consensus/Types/EpochDep
  - use `bsiLIWS` (instead of `bootstrapQC`)

NOTE: There are still places that need to be updated to use the InitProperties contract.
      I.E., Common and VotesOnce.
      That will be done in a later PR.

### CODE REVIEW

File: LibraBFT/Impl/Handle.agda [hunk 1]
```diff
@@ -32,157 +35,148 @@ open import LibraBFT.Yasm.Base
...
+  That functions calls 'ConsensusProvider.startConsensus' which returns
+  '(EpochManager, [Output]'.
+
+  'Start.startViaConsensusProvider' then goes on to create and wire up internal communication
+  channels and starts threads.  The most relevant thread starts up
+  `(EpochManager.obmStartLoop epochManager output ...)' to handle the initialization output
+  and then to handle new messages from the network.
+
+  In Agda functions below, assuming no initialization errors,
+  - the 'GenKeyFile.create' function
```
Comment (mark-moir) [inline LibraBFT/Impl/Handle.agda:79]: I think these comments are out of date, as the Agda code does not create the file and read it to construct `BootStrapInfo`; we simply assume `bootStrapInfo` known to everyone.  I think it's good to keep the comments about how the Haskell implementation initialises, but also to be accurate about how we're approaching it in Agda.

File: LibraBFT/Impl/Handle.agda [hunk 2]
```diff
@@ -32,157 +35,148 @@ open import LibraBFT.Yasm.Base
...
+  abstract
+    initialize  : Instant → GenKeyFile.NfLiwsVsVvPe → Either ErrLog (EpochManager × List Output)
+    initialize  = initialize'
+    initialize≡ : initialize ≡ initialize'
+    initialize≡ = refl
+
+  mkNfLiwsVsVvPe : BootstrapInfo → ValidatorSigner → GenKeyFile.NfLiwsVsVvPe
+  mkNfLiwsVsVvPe bsi vs = (bsi ^∙ bsiNumFaults , bsi ^∙ bsiLIWS , vs , bsi ^∙ bsiVV , bsi ^∙ bsiPE)
+
+  initEMWithOutput' : BootstrapInfo → ValidatorSigner → Either  ErrLog (EpochManager × List Output)
```
Comment (mark-moir) [inline LibraBFT/Impl/Handle.agda:111]: As there is no branching here, it probably won't be necessary/beneficial to have an `EitherD` variant, but it doesn't do much harm to keep it.

File: LibraBFT/Impl/Properties/VotesOnce.agda [hunk 3]
```diff
@@ -53,7 +54,7 @@ newVote⇒lv≡
     → Meta-Honest-PK pk → ¬ (∈BootstrapInfo-impl fakeBootstrapInfo (ver-signature sig))
     → ¬ MsgWithSig∈ pk (ver-signature sig) (msgPool pre)
     → LastVoteIs s' v
-newVote⇒lv≡ _ (step-init initSucc ini) _ send∈acts = ⊥-elim (obm-dangerous-magic' "The Contract for the init handler should say that it sends no messages, contradicting send∈acts")
+newVote⇒lv≡ _ (step-init initSucc uni) _ send∈acts = ⊥-elim (obm-dangerous-magic' "The Contract for the init handler should say that it sends no messages, contradicting send∈acts")
```
Comment (mark-moir) [inline LibraBFT/Impl/Properties/VotesOnce.agda:57]: You've pointed out that this comment is wrong.  I think you fixed it somewhere else, but please fix this one too.

File: LibraBFT/ImplShared/Consensus/Types.agda [hunk 4]
```diff
@@ -185,18 +185,25 @@ module LibraBFT.ImplShared.Consensus.Types where
   record BootstrapInfo : Set where
     constructor mkBootstrapInfo
     field
-      -- TODO-1 : Nodes, PKs for initial epoch
-      -- TODO-1 : Faults to tolerate (or quorum size?)
-      bootstrapQC : QuorumCert -- use the same bootstrap QC for highestQC and highestCommitCert
-  open BootstrapInfo
+      _bsiNumFaults : ℕ
+      _bsiLIWS      : LedgerInfoWithSignatures
+      _bsiVSS       : List ValidatorSigner -- TODO-1 : remove this after ValidatorSigner
```
Comment (mark-moir) [inline LibraBFT/ImplShared/Consensus/Types.agda:190]: This comment is unclear.  What is "this"?  Remove it "after" what?

**Review by mark-moir — APPROVED** [review]
@haroldcarr woohoo, great progress!  So much real, so much less fake :smile:.  After you address my small number of (mostly meta) comments, APPROVED.  Thanks.

File: LibraBFT/Impl/Handle.agda [hunk 5]
```diff
@@ -32,157 +35,148 @@ open import LibraBFT.Yasm.Base
...
+
+  InitAndHandlers : SystemInitAndHandlers ℓ-RoundManager ConcSysParms
+  InitAndHandlers =
+    mkSysInitAndHandlers
+      fakeBootstrapInfo
+      fakeInitRM
+      initHandler
+      peerStep
+   where
+    postulate fakeInitRM : RoundManager
```
[the remark this answers is not in the document]
Comment (haroldcarr) [reply LibraBFT/Impl/Handle.agda:182]: I changed it to this:

```
  InitAndHandlers : SystemInitAndHandlers ℓ-RoundManager ConcSysParms
  InitAndHandlers =
    mkSysInitAndHandlers
      fakeBootstrapInfo
      fakeInitRM
      initHandler
      peerStep
   where
    postulate
      bs : BlockStore
      pe : ProposerElection
      rs : RoundState
      sr : SafetyRules
      vv : BootstrapInfo → ValidatorVerifier
    -- For uninitialised peers, so we know nothing about their state.
    -- Construct a value of type `RoundManager` to ensure it is inhabitable.
    fakeInitRM : RoundManager
    fakeInitRM = RoundManager∙new
      ObmNeedFetch∙new
      (EpochState∙new 1 (vv fakeBootstrapInfo))
      bs rs pe proposalGenerator sr false
```

File: LibraBFT/Impl/Handle.agda (hunk 1 above)
Comment (haroldcarr) [reply LibraBFT/Impl/Handle.agda:79]: DONE

File: LibraBFT/Impl/Handle.agda (hunk 2 above)
Comment (haroldcarr) [reply LibraBFT/Impl/Handle.agda:111]: I'll leave it for now.

File: LibraBFT/Impl/Properties/VotesOnce.agda (hunk 3 above)
Comment (haroldcarr) [reply LibraBFT/Impl/Properties/VotesOnce.agda:57]: DONE. I updated this and several other places appropriately.

File: LibraBFT/ImplShared/Consensus/Types.agda (hunk 4 above)
Comment (haroldcarr) [reply LibraBFT/ImplShared/Consensus/Types.agda:190]: I removed the comment.
What it is talking about no longer exists.

**Pull request merged**